### ISSUE TITLE
potential-bugs Deprecation does not work

### ISSUE THREAD

**Issue by kepocnhh** [arturbosch/detekt#2038]
```kotlin
fun deprecationSample() {
    val r = 1 in 0.0..5.0
    Date(2019, 5, 5)
}
```

- Operator `in` is deprecated in kotlin because "This `contains` operation mixing integer and floating point arguments has ambiguous semantics and is going to be removed."
- Constructor `Date` is deprecated in java because replaced by `Calendar`

## Expected Behavior
The rule will work and the detekt will report an error.

## Observed Behavior
Detekt does not report an error.

## Steps to Reproduce
I use my task:
```kotlin
task<Detekt>("verifyQuality") {
    source = files(subprojects.withPlugin("kotlin").srcDirs("main")).asFileTree
    config = files(
        "buildSrc/src/main/resources/detekt/config/potential_bugs.yml"
    )
    reports {
        html {
            enabled = true
            destination = File(analysisQualityHtmlPath)
        }
        xml.enabled = false
        txt.enabled = false
    }
}
```

File potential_bugs.yml contains:
```yaml
build:
  maxIssues: 0

potential-bugs:
  active: true
  Deprecation:
    active: true

processors:
  active: false
```

## Context
Detekt definitely works on other rule sets (for example, DuplicateCaseInWhenExpression). That is, the problem should not be in the source of the source code or the path to detekt configuration file. For some reason I can’t turn on this particular rule (Deprecation).

## Environment
java 1.8.0_221
kotlin 1.3.50
detekt 1.1.1
Gradle 5.5.1
macOS Mojave Version 10.14.6 (18G87)

**Comment by arturbosch** [comment]
Thanks for the great issue report.
`Deprecation` is a rule which relies on type resolution of the kotlin compiler which is experimental for now and needs two extra task properties `classpath` and `jvmTarget`.
We do not even ship documentation for them for now, sorry.
The Cli documents the flags when `--help` is used: https://github.com/arturbosch/detekt/blob/e3ab2f5454d4b60b45eda706ae17dcc36ea84682/detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/CliArgs.kt#174

**Comment by arturbosch** [comment]
@schalkms @3flex I labeled and closed all these issues so we can track them after 1.2

**Comment by schalkms** [comment]
We now have docs for these issues. Can we close this issue?

**Comment by kepocnhh** [comment]
Can you help me find where I can see information about this?

**Comment by schalkms** [comment]
Yes sure.
The [README page](https://github.com/arturbosch/detekt#with-gradle) shows how to configure type resolution in Gradle.
Furthermore, I created the PR #2198 that also adds this to the homepage.

**Comment by kepocnhh** [comment]
So. I updated Kotlin to `1.3.61` and Detect to `1.3.0`. I added the lines:
```kotlin
tasks {
    withType<Detekt> {
        // Target version of the generated JVM bytecode. It is used for type resolution.
        this.jvmTarget = "1.8"
    }
}
```
as indicated in the readme. But the `Deprecation: active: true` check still doesn't work. Please help me figure out what I'm missing.

**Comment by schalkms** [comment]
Okay. Do the other rules in the `bugs` ruleset that use type resolution work for you? Can you please check that @kepocnhh ?

* MissingWhenCase
* HasPlatformType
* UnsafeCallOnNullableType
* UnsafeCast
* RedundantElseInWhen

**Comment by kepocnhh** [comment]
```kotlin
private enum class E {
    E1,
    E2
}
private fun test() {
    //Deprecation
    Date(2019, 5, 5)
    //MissingWhenCase
    val e: E? = null
    when(e) {
        E.E1 -> Unit
    }
    //HasPlatformType
    val s = System.getProperty("key")
    //UnsafeCallOnNullableType
    val someEnum: E? = E.values().firstOrNull()
    println(someEnum!!.name)
    //UnsafeCast
    val a: Any = E.values().first().ordinal
    println(a as String)
    println(a as E)
    //RedundantElseInWhen
    when (e) {
        E.E1 -> Unit
        E.E2 -> Unit
        null -> Unit
        else -> Unit
    }
}
```

No changes so far. I apparently missed something. I can’t understand what I'm doing wrong.

**Comment by schalkms** [comment]
So, does detekt flag these issues with `active: true`?

**Comment by kepocnhh** [comment]
Yes. File potential_bugs.yml https://gist.github.com/kepocnhh/525086cfb56b73e7e46d38a2bf4353a6

as before
```kotlin
task<Detekt>("verifyQuality") {
    source = files(subprojects.withPlugin("kotlin").srcDirs("main")).asFileTree
    config = files(
        "buildSrc/src/main/resources/detekt/config/potential_bugs.yml"
    )
    reports {
        html {
            enabled = true
            destination = File(analysisQualityHtmlPath)
        }
        xml.enabled = false
        txt.enabled = false
    }
}
```
but now with
```kotlin
tasks {
    withType<Detekt> {
        // Target version of the generated JVM bytecode. It is used for type resolution.
        this.jvmTarget = "1.8"
    }
}
```

**Comment by schalkms** [comment]
If I understood you correctly, the problem is concerning only the `Deprecation` rule. All the other rules work fine. This is weird.
Did you configure the `classpath` property, which should point to the class and depending jar files?

[CLI Doc](https://arturbosch.github.io/detekt/cli.html)

**Comment by kepocnhh** [comment]
- Deprecation 
- MissingWhenCase
- HasPlatformType
- UnsafeCallOnNullableType
- UnsafeCast
- RedundantElseInWhen

All these rules do not work.
And I do not configure the classpath property. I use the detekt only from the gradle task, as described above. The link that you gave me says how to run detekt using Command Line Interface. Please explain how this should help me to use the detekt from the gradle task.

**Comment by schalkms** [comment]
> All these rules do not work.

That's an important piece of information. There's no problem with detekt (or its rules) but rather with the wrong user configuration, which differs from the recommended way on our [homepage](https://arturbosch.github.io/detekt/kotlindsl.html).

> And I do not configure the classpath property.

In order to use the mentioned rules you have to configure the `classpath` property inside one of the tasks. The mentioned rules do advanced static analysis (type resolution). For that reason the rules need a path that points to the class and depending jar files. Otherwise, the rules cannot be used.

>  The link that you gave me says how to run detekt using Command Line Interface. 

The link gives an overview of all the available config options. It shows the description for the `classpath` option.

> `source = files(subprojects.withPlugin("kotlin").srcDirs("main")).asFileTree`

I double-checked your config. Why do you use `source` and not `input` here?

**Comment by kepocnhh** [comment]
> Why do you use source and not input here?

Detekt `1.1.1` `@Deprecated("Replace with getSource/setSource")`
Detekt `1.3.0` does not have property `input`
Correct me please if I'm wrong.

> The link gives an overview of all the available config options. It shows the description for the classpath option.

[Detekt.kt](https://github.com/arturbosch/detekt/blob/master/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt) has `detektClasspath`, `pluginClasspath` and `classpath` but final. I still do not understand how I can set these parameters.

Perhaps the description at https://arturbosch.github.io/detekt/kotlindsl.html is out of date?

@arturbosch @schalkms @3flex

**Comment by schalkms** [comment]
> Perhaps the description at https://arturbosch.github.io/detekt/kotlindsl.html is out of date?

It is, yes. I have one last question. Is the shown config on the homepage (`detekt { }` instead of a `custom task "verifyQuality"`) working for you if you replace `input` with `source`?

**Comment by kepocnhh** [comment]
I used the configuration from the site.

```kotlin
detekt {
    toolVersion = "1.3.0"

    input = files(subprojects.withPlugin("kotlin").srcDirs("main"))
    val configPath = "buildSrc/src/main/resources/detekt/config"
    config.setFrom(
        "$configPath/common.yml",
        "$configPath/potential_bugs.yml"
    )
    reports {
        html {
            enabled = true
            destination = File(analysisQualityHtmlPath)
        }
        xml.enabled = false
        txt.enabled = false
    }
}
```

Indeed, `io.gitlab.arturbosch.detekt.extensions.DetektExtension` has an `input` field (but no `source`), and in `io.gitlab.arturbosch.detekt.Detekt` `input` has already been replaced with `source`.

But so far I have the same result. `Deprecation` does not work. `DuplicateCaseInWhenExpression` work.

> Are you using detekt in a multi-module project?

Yes. I describe the task in the root module (build.gradle.kts). I analyze the code in submodules.
- build.gradle.kts
- lib1/src/kotlin...
- lib2/src/kotlin...
- ...

Therefore, I made a one-module project with configuration from the site.
- build.gradle.kts
- src/kotlin...

Same config. Same code. Add lines
```kotlin
tasks {
    withType<Detekt> {
        this.jvmTarget = "1.8"
    }
}
```
But so far I have the same result. `Deprecation` does not work. `DuplicateCaseInWhenExpression` work.

Please help determine what I can miss.

**Comment by schalkms** [comment]
> Therefore, I made a one-module project with configuration from the site.

Could you please share this project in order to be able to reproduce it?

I have one more questions in order to debug the problem.
Do you use detekt in an Android project?

**Comment by kepocnhh** [comment]
> Do you use detekt in an Android project?

Not yet, but planning.

> Could you please share this project in order to be able to reproduce it?

https://github.com/kepocnhh/DetektSample

Just run the `gradle detekt` first in directory `singlemodule`, and then in directory `multimodule`. In my case, the `detekt` does not respond to code in the `detekt.sample.App.kt`.

## Environment
java openjdk 12.0.2
kotlin 1.3.61
detekt 1.4.0
Gradle 5.5.2
Microsoft Windows 10 Pro N 10.0.18363

**Comment by schalkms** [comment]
Thanks for helping us out. I can reproduce it with your sample project.
@3flex @arturbosch do you have any idea why the classpath isn't picked up and hence the rules using type solving aren't working?

**Comment by arturbosch** [comment]
@kepocnhh if you run `detekt` in `singlemodule` you won't get any findings reported. But if you run `detektMain`, you get:
```
/home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt - 1h 5min debt
	Deprecation - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:16:5
	MissingWhenCase - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:19:5
	RedundantElseInWhen - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:32:5
	UnsafeCallOnNullableType - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:26:13

Overall debt: 1h 5min

Ruleset: potential-bugs - 1h 5min debt
	Deprecation - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:16:5
	MissingWhenCase - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:19:5
	RedundantElseInWhen - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:32:5
	UnsafeCallOnNullableType - [test] at /home/artur/git/test/DetektSample/singlemodule/src/main/kotlin/detekt/sample/App.kt:26:13

Overall debt: 1h 5min

detekt finished in 2375 ms.
Successfully generated HTML report at /home/artur/git/test/DetektSample/singlemodule/build/reports/detekt/main.html
Successfully generated Checkstyle XML report at /home/artur/git/test/DetektSample/singlemodule/build/reports/detekt/main.xml
Successfully generated plain text report at /home/artur/git/test/DetektSample/singlemodule/build/reports/detekt/main.txt
Build failed with 4 weighted issues (threshold defined was 0).
io.gitlab.arturbosch.detekt.cli.BuildFailure: Build failed with 4 weighted issues (threshold defined was 0).
```

This is because the type resolution is experimental and only configured for detekt[SourceSet]. Only custom tasks are allowed to override the `classpath` property. The `DetektExtension` does not have it, so it won't get passed as an input parameter to the `detekt-cli`.

If you are sure about using these rules now, you have to use `detektMain` and pay the extra performance costs. Feel free to follow https://github.com/detekt/detekt-compiler-plugin for progress on running detekt as a compiler plugin without performance loss (not yet for production).

**Comment by schalkms** [comment]
@kepocnhh thanks for your patience while debugging and solving your problem. We try to improve the documentation as described in #2259.
I'll close this issue now. If you still experience problems, please feel free to comment here or open a new issue.
